File: Classic Algorithms/Valid Parenthesis.py


def isValid(s):
    stack = []

    for item in s:

        if item in ["(", "{", "["]:

            stack.append(item);

        elif len(stack) == 0:

            return False

        else:

            poppedItem = stack.pop()

            if item == ")" and poppedItem != "(":
                return False

            if item == "}" and poppedItem != "{":
                return False

            if item == "]" and poppedItem != "[":
                return False

    else:

        if len(stack) == 0:
            return True
        else:
            return False


# Driver Code

# True
result = isValid("([])")
print(result)

# False
result = isValid("(((")
print(result)

# True
result = isValid("(){}[]")
print(result)

# False
result = isValid("({}([]")
print(result)

File: Classic Algorithms/Valid Parenthesis.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = s => {
    
    stack = [];
    
    for (let item of s) {
        if (item === "(" || item === "{" || item === "[") {
            stack.push(item);
        }
        else if (stack.length === 0) {
            return false;
        }
        else {
            poppedItem = stack.pop();
            
            if (item === ")" && poppedItem !== "(") {
                return false; 
            }
            
            if (item === "}" && poppedItem !== "{") {
                return false;
            }
            
            if (item === "]" && poppedItem !== "[") {
                return false;
            }
        }
    }
    
    return stack.length > 0 ? false : true;
};